cmd/snap-bootstrap/initramfs_install_handler.go

```diff
@@ -0,0 +1,418 @@
+)
+
+var (
+	// assertion store to be trusted
+	trusted = sysdb.Trusted()
+	// /run/mnt/ubuntu-data is not exported, why?
+	InstallUbuntuDataDir = filepath.Join(boot.InitramfsRunMntDir, "ubuntu-data")
+)
+
+// mounts info
+type mountInfo struct {
+	mountPoint string
+	unmountFnc func() error
+	// essential flag is tracking mounts essential for install mode
+	// those mounts should be destroyed when moving to run mode
+	essential bool
+}
+
+// tracking of active mounts
+var activeMounts []mountInfo
+
+func doUnmount(info *mountInfo) error {
+	if err := info.unmountFnc(); err != nil {
+		// only log error and try to unmount next
+		return fmt.Errorf("Failed to unmount%s, %v", info.mountPoint, err)
+	}
+	return nil
+}
+
+func hasFDESetupHook(kernelInfo *snap.Info) (bool, error) {
+	_, err := exec.LookPath(InitRamfsFdeHook)
+	return err == nil, err
+}
+
+func runFDESetupHook(req *fde.SetupRequest) ([]byte, error) {
+	stdin, err := json.Marshal(req)
+	if err != nil {
+		return nil, fmt.Errorf(`cannot build request %v for fde-reveal-key: %v`, req, err)
+	}
+	return fde.RunFDEinitramfsHelper(InitRamfsFdeHook, stdin)
+}
+
+func newMemAssertionsDB(commitObserve func(verified asserts.Assertion)) (db *asserts.Database, commitTo func(*asserts.Batch) error, err error) {
+	memDB, err := asserts.OpenDatabase(&asserts.DatabaseConfig{
+		Backstore: asserts.NewMemoryBackstore(),
+		Trusted:   trusted,
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	commitTo = func(b *asserts.Batch) error {
+		return b.CommitToAndObserve(memDB, commitObserve, nil)
+	}
+
+	return memDB, commitTo, nil
+}
+
+func loadDeviceSeed(systemLabel string) (deviceSeed seed.Seed, err error) {
+	deviceSeed, err = seed.Open(boot.InitramfsUbuntuSeedDir, systemLabel)
+	if err != nil {
+		return nil, err
+	}
+	n := runtime.NumCPU()
+	if n > 1 {
+		deviceSeed.SetParallelism(n)
+	}
+	return deviceSeed, err
+}
+
+func loadSeedAssertions(deviceSeed seed.Seed) (*asserts.Database, error) {
+	start := time.Now()
+	db, commitTo, err := newMemAssertionsDB(nil)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to open temp database: %v\n", err)
+	}
+	if err = deviceSeed.LoadAssertions(db, commitTo); err != nil {
+		return nil, fmt.Errorf("Failed to load assertions: %v\n", err)
+	}
+	logger.Noticef("Seed assertions loaded in %s\n", time.Since(start))
+	return db, nil
+}
+
+func loadSeedEssentialMeta(deviceSeed seed.Seed, sysLabel string) error {
+	perfTimings := timings.New(nil)
+	// measure meta loading
+	start := time.Now()
+	logger.Noticef("Starting to load Seed Meta\n")
+	snap.SanitizePlugsSlots = builtin.SanitizePlugsSlots
+
+	if err := deviceSeed.LoadEssentialMeta([]snap.Type{snap.TypeGadget, snap.TypeBase, snap.TypeKernel}, perfTimings); err != nil {
+		return fmt.Errorf("Failed to load essential seed meta: %v\n", err)
+	}
+
+	logger.Noticef("Seed meta loaded in %s\n", time.Since(start))
+	return nil
+}
+
+// helper to check if slice contains
+// TODO: remove when moving to go 1.18
```

Comment:
are you thinking of this https://pkg.go.dev/golang.org/x/exp/slices ? not sure we will start using that extensively